### PR TITLE
ALB support

### PR DESCRIPTION
**Pull request by dpetzold**
Support creating an ALB stack with target groups, listeners, security groups and alarms. A sample stack is defined below:

```
- class_path: stacker_blueprints.alb.LoadBalancer
  description: Application load balancer for ${service}
  name: ${service}-load-balancer
  variables:
    Name: ${namespace}-${service}
    Scheme: internal
    Type: application
    IpAddressType: ipv4
    TargetGroups:
    - Name: ${namespace}-${service}-http
      HealthCheck:
        Port: 443
        Protocol: https
        IntervalSeconds: 10
        Path: /healthcheck
        TimeoutSeconds: 5
        HealthyThresholdCount: 5
        UnhealthyThresholdCount: 2
        SuccessCodes:
          - 200
      Port: 80
      Protocol: http
      TargetType: instance
    - Name: ${namespace}-${service}-https
      HealthCheck:
        IntervalSeconds: 10
        Path: /healthcheck
        TimeoutSeconds: 5
        HealthyThresholdCount: 5
        UnhealthyThresholdCount: 2
        SuccessCodes:
          - 200
      Port: 443
      Protocol: https
      TargetType: instance
      Alarms:
      - ActionsEnabled: true
        AlarmActions:
          - ${alarm_sqs}
        AlarmDescription: Unhealthy hosts
        ComparisonOperator:  GreaterThanOrEqualToThreshold
        EvaluationPeriods: 1
        MetricName: UnHealthyHostCount
        Period: 300
        Statistic: Sum
        Threshold: "1"
    Listeners:
    - DefaultActions:
      - TargetGroup: ${namespace}-${service}-http
        Type: forward
      Port: 80
      Protocol: HTTP
    - Certificates:
        - ${certificate_arn}
      DefaultActions:
      - TargetGroup: ${namespace}-${service}-https
        Type: forward
      Port: 443
      Protocol: HTTPS
    EgressRules:
    - CidrIp: 0.0.0.0/0
      FromPort: 0
      ToPort: 0
      IpProtocol: "-1"
    IngressRules:
    - CidrIp: 0.0.0.0/0
      FromPort: 80
      ToPort: 80
      IpProtocol: tcp
    - CidrIp: 0.0.0.0/0
      FromPort: 443
      ToPort: 443
      IpProtocol: tcp
    SubnetIds: ${private_subnets}
    VpcId: ${vpc_id}
    Alarms:
    - ActionsEnabled: true
      AlarmActions:
        - ${alarm_sqs}
      AlarmDescription: 500 error rate exceeded
      ComparisonOperator:  GreaterThanOrEqualToThreshold
      EvaluationPeriods: 1
      MetricName: HTTPCode_Target_5XX_Count
      Period: 300
      Statistic: Average
      Threshold: "350"
    - ActionsEnabled: true
      AlarmActions:
        - ${alarm_sqs}
      AlarmDescription: 400 error rate exceeded
      ComparisonOperator:  GreaterThanOrEqualToThreshold
      EvaluationPeriods: 1
      MetricName: HTTPCode_Target_4XX_Count
      Period: 300
      Statistic: Average
      Threshold: "500"
```

### CODE REVIEW

File: stacker_blueprints/alb.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+            alarm['Dimensions'] = [
+                {
+                    'Name': 'LoadBalancer',
+                    'Value': load_balancer.GetAtt('LoadBalancerFullName'),
+                },
+            ]
+
+            self.create_alarm(alarm_name, alarm)
+
+        t.add_output(
```
Comment (phobologic) [inline stacker_blueprints/alb.py]: A small update to stacker recently gave a helper that might be good to use in here - the blueprint base class in 1.4 has an `add_output` helper:

```
self.add_output("LoadBalancerArn", load_balancer.Ref())
```

File: stacker_blueprints/alb.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+            hc_attrs = {
+                'HealthCheckIntervalSeconds': hc['IntervalSeconds'],
+                'HealthCheckPath': hc['Path'],
+                'HealthCheckTimeoutSeconds': hc['TimeoutSeconds'],
+                'HealthyThresholdCount': hc['HealthyThresholdCount'],
+                'UnhealthyThresholdCount': hc['UnhealthyThresholdCount'],
+            }
+
+            hc_port = hc.get('Port')
```
Comment (phobologic) [inline stacker_blueprints/alb.py]: This bit of logic can be simplified a little with:

```
hc_attrs["HealthCheckPort"] = hc.get("Port", NoValue)
```

File: stacker_blueprints/alb.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+            tags = tg.get('Tags', {})
+            tags.setdefault('Name', tg['Name'])
+
+            target_group = t.add_resource(
+                alb.TargetGroup(
+                    logical_name,
+                    Matcher=alb.Matcher(
+                        HttpCode=','.join(str(sc) for sc in hc['SuccessCodes'])
+                    ),
+                    Name=tg['Name'],
```
Comment (phobologic) [inline stacker_blueprints/alb.py]: Again I'd remove the name

File: stacker_blueprints/alb.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+            for attr in ['TargetGroupFullName', 'TargetGroupName']:
+                t.add_output(
+                    Output(logical_name + attr, Value=target_group.GetAtt(attr))
+                )
+
+            target_groups[tg['Name']] = target_group
+
+            alarms = tg.get("Alarms", [])
+            for alarm in alarms:
```
Comment (phobologic) [inline stacker_blueprints/alb.py]: Can you break this off into a separate method called `create_target_alarms`?

File: tests/test_alb.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
+from stacker.context import Context
```
Comment (phobologic) [inline tests/test_alb.py]: Can you convert this to the new YAML style tests?  It's actually a lot easier to write, and can serve as really good examples.  See https://github.com/cloudtools/stacker_blueprints/blob/master/tests/test_ecs.yaml for an example.

**Comment by dpetzold** [comment]
@phobologic thanks for your feedback. I think I covered all the points expect I made "Name" optional. I agree its best practice not to use it but there are cases where people may want to.

**Comment by phobologic** [comment]
I'd still rather not have the Name - especially for a service like a LoadBalancer, where completely replacing it can result in signficant downtime.  The description is helpful, but I'd really rather push for safer standards in the standard blueprints/examples, to give people a good standard to work with.  My nightmare is someone using one of these blueprints and ending up in a situation where an update will require downtime.

**Pull request closed without merging**